async_downloads/static/js/ws_async_downloads.js

```diff
@@ -0,0 +1,198 @@
+    const download_div = `
+    <div id="${data.downloadKey}" class="ws-download-container" data-url="${download.url}"
+        data-filepath="${download.filepath}">
+        <div class="ws-product-label background-color-${download.product}"></div>
+        <div class="download-content">
+            <div class="download">
+                <span class="download-name">${download.name}</span>
+                <span class="download-timestamp" value="${download.timestamp}">now</span>
+            </div>
+            <span class="download-clear"><i class="fa fa-times-circle"></i></span>
+
+            <div class="progress" style="height: 4px;">
+                <div class="progress-bar progress-bar-striped bg-info progress-bar-animated" role="progressbar"
+                    style="width: 0%" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100"></div>
+            </div>
+        </div>
+    </div>
+    `;
+    dropdown.prepend(download_div);
+}
+
+function updateDownload(data) {
+    const download_div = $(`#${data.downloadKey}`);
+    const download = data.download;
+    const bar = download_div.find(".progress-bar");
+    download_div.attr("data-url", download.url);
+    bar.attr({
+        style: `width: ${download.percentage}%`,
+        "aria-valuenow": download.percentage,
+    });
+    if (download.complete) {
+        bar.addClass("bg-success");
+        bar.removeClass("bg-info progress-bar-animated");
+    }
+    if (download.errors) {
+        bar.addClass("bg-danger");
+        bar.removeClass("bg-success");
+        download_div.attr({
+            "data-toggle": "tooltip",
+            "data-placement": "bottom",
+            title: download.errors,
+        });
+        $("#async-downloads [data-toggle='tooltip']").tooltip();
+    }
+}
+function removeDownload(data) {
+    $(`#${data.downloadKey}`).remove();
+    if ($("#downloads-dropdown > div").length === 0)
+        $("#downloads-dropdown").append(
+            '<a class="dropdown-item" href="">No downloads</a>'
+        );
+}
+
+function startDownloadSocket() {
+    const master_stripped_url = JSON.parse(
+        document.getElementById("MASTER_URL").textContent
+    ).replace(/(^\w+:|^)\/\//, "");
+    const username = JSON.parse(
+        document.getElementById("username").textContent
+    );
+    const ws_scheme = window.location.protocol == "https:" ? "wss" : "ws";
+
+    const ws = new WebSocket(
+        `${ws_scheme}://${master_stripped_url}/ws/downloads/${username}/`
+    );
+
+    function requestDownloadUpdate(data) {
+        setTimeout(function () {
+            const downloadKey = data.downloadKey;
+            if (!data.download.complete) {
+                ws.send(
+                    JSON.stringify({
+                        eventType: "getDownload",
+                        location: window.location.host,
+                        downloadKey: downloadKey,
+                    })
+                );
+            }
+        }, 1000);
```

Comment:
If we are using websockets, why are we using `setTimeout` here? This looks a bit like polling.

Reply:
We pooling here download update each second, to the moment when it is completed via websockets. First I tried to create that in backed but due to strange bug or something that I was not able to resolve, functionality was moved into fronted.

I tried in many places inside consumer to create endless loop with await to send notification for all channels. But this `channels` library works in some weird way and even if we are awaiting for function inside `group_send()` all the messages are cummulated and then executed at once via events :/

Reply:
We will need to have the `update_percentage` function be able to send a websocket message rather than polling on the front-end.

Reply:
Changed here: 3b316ac
![Peek 2022-01-25 14-03](https://user-images.githubusercontent.com/37776903/150982121-2f5eed7c-0ac6-4508-b658-c379137cae8e.gif)